src/libraries/System.Net.NameResolution/tests/FunctionalTests/GetHostAddressesTest.cs

```diff
@@ -118,5 +139,43 @@ public void DnsGetHostAddresses_LocalHost_ReturnsSameAsGetHostEntry()
 
             Assert.Equal(ipEntry.AddressList, addresses);
         }
+
+        [Theory]
+        [InlineData(false, AddressFamily.InterNetwork)]
+        [InlineData(false, AddressFamily.InterNetworkV6)]
+        [InlineData(true, AddressFamily.InterNetwork)]
+        [InlineData(true, AddressFamily.InterNetworkV6)]
+        public void DnsGetHostAddresses_LocalHost_AddressFamilySpecific(bool useAsync, AddressFamily addressFamily)
+        {
+            IPAddress[] addresses =
+                useAsync ? Dns.GetHostAddressesAsync(TestSettings.LocalHost, addressFamily).GetAwaiter().GetResult() :
+                Dns.GetHostAddresses(TestSettings.LocalHost, addressFamily);
+
+            Assert.All(addresses, address => Assert.Equal(addressFamily, address.AddressFamily));
+        }
+
+        [Fact]
+        public async Task DnsGetHostAddresses_PreCancelledToken_Throws()
+        {
+            using var cts = new CancellationTokenSource();
+            cts.Cancel();
+
+            OperationCanceledException oce = await Assert.ThrowsAnyAsync<OperationCanceledException>(() => Dns.GetHostAddressesAsync(TestSettings.LocalHost, cts.Token));
+            Assert.Equal(cts.Token, oce.CancellationToken);
+        }
+
+        [Fact]
+        public async Task DnsGetHostAddresses_PostCancelledToken_Throws()
+        {
+            using var cts = new CancellationTokenSource();
+
+            Task task = Dns.GetHostAddressesAsync(TestSettings.UncachedHost, cts.Token);
+
+            // This test might flake if the cancellation token takes too long to trigger.
```

Comment:
How can we avoid that?  At a minimum it sounds like the test should be made outerloop, but even beyond that, we should avoid introducing flaky tests.  Should we add retries?  Is there something we can do to make it not-flaky by construction?  etc.

Reply:
The race is between calling GetAddrInfoExCancel and getting a reply... so perhaps we can buy a domain and point it to a slow to reply, or even non-existent, DNS server.

Reply:
This is an example of a test where we might consider running in a container with DNS name servers defined as something we control (i.e. another container running DNS) etc.